pkg/cluster/gossip.go

```diff
@@ -0,0 +1,253 @@
+		return fmt.Errorf("at most one of join peers and discover peers may be set")
+	} else if c.DiscoverPeers != "" {
+		providers := make(map[string]discover.Provider, len(discover.Providers)+1)
+		for k, v := range discover.Providers {
+			providers[k] = v
+		}
+		// Extra providers used by tests
+		for k, v := range extraDiscoverProviders {
+			providers[k] = v
+		}
+
+		// Custom providers that aren't enabled by default
+		providers["k8s"] = &k8s.Provider{}
+
+		d, err := discover.New(discover.WithProviders(providers))
+		if err != nil {
+			return fmt.Errorf("bootstrapping peer discovery: %w", err)
+		}
+
+		addrs, err := d.Addrs(c.DiscoverPeers, stdlog.New(io.Discard, "", 0)) // TODO(rfratto): log to log.Logger?
+		if err != nil {
+			return fmt.Errorf("discovering peers: %w", err)
+		}
+		c.JoinPeers = addrs
+	}
+
+	for i := range c.JoinPeers {
+		// Default to using the same advertise port as the local node. This may
+		// break in some cases, so the user should make sure the port numbers
+		// align on as much nodes as possible.
+		c.JoinPeers[i] = appendDefaultPort(c.JoinPeers[i], defaultPort)
+	}
+
+	return nil
+}
+
+func appendDefaultPort(addr string, port int) string {
+	_, _, err := net.SplitHostPort(addr)
+	if err == nil {
+		// No error means there was a port in the string
+		return addr
+	}
+	return fmt.Sprintf("%s:%d", addr, port)
+}
+
+// GossipNode is a Node which uses gRPC and gossip to discover peers.
+type GossipNode struct {
+	// NOTE(rfratto): GossipNode is a *very* thin wrapper over ckit.Node, but it
+	// still abstracted out as its own type to have more agent-specific control
+	// over the exposed API.
+
+	cfg       *GossipConfig
+	innerNode *ckit.Node
+	log       log.Logger
+	sharder   shard.Sharder
+
+	started atomic.Bool
+}
+
+// NewGossipNode creates an unstarted GossipNode. The GossipNode will register
+// itself as a gRPC service to srv. GossipConfig is expected to be valid and
+// have already had ApplyDefaults called on it.
+//
+// GossipNode operations are unavailable until the node is started.
+func NewGossipNode(l log.Logger, srv *grpc.Server, c *GossipConfig) (*GossipNode, error) {
+	if l == nil {
+		l = log.NewNopLogger()
+	}
+
+	sharder := shard.Ring(256)
+
+	ckitConfig := ckit.Config{
+		Name:          c.NodeName,
+		AdvertiseAddr: c.AdvertiseAddr,
+		Sharder:       sharder,
+		Log:           l,
+		Pool:          c.Pool,
+	}
+
+	inner, err := ckit.NewNode(srv, ckitConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	return &GossipNode{
+		cfg:       c,
+		innerNode: inner,
+		log:       l,
+		sharder:   sharder,
+	}, nil
+}
+
+// ChangeState changes the state of n. ChangeState will block until the state
+// change has been receieved by another node; cancel the context to stop
+// waiting.
+//
+// Nodes must be a StateParticipant to be returned during Lookup.
+//
+// The following state transitions are valid:
+//
+// 		StatePending -> StateParticipant|StateViewer|StateTerminating|StateGone
```

Comment:
Does StatePending exist? StateViewer seems to be the default state

Reply:
Stale comment, I need to update this or stop trying to sync it with the upstream comment

Reply:
(but yes, StatePending used to exist and was since removed)